File: 0130-surrounded-regions/0130-surrounded-regions.py
from collections import deque


class Solution:
    def solve(self, board: List[List[str]]) -> None:
        if not board or not board[0]: return
        row, col = len(board), len(board[0])
        if row < 3 or col < 3: return

        dr = [0, 0, -1, 1]
        dc = [1, -1, 0, 0]
        chk = [[False for i in range(col)] for j in range(row)]
        que = deque()

        for i in range(row):
            if board[i][0] == 'O':
                chk[i][0] = True
                que.append((i, 0))
            if board[i][col - 1] == 'O':
                chk[i][col - 1] = True
                que.append((i, col - 1))

        for i in range(col):
            if board[0][i] == 'O':
                chk[0][i] = True
                que.append((0, i))
            if board[row - 1][i] == 'O':
                chk[row - 1][i] = True
                que.append((row - 1, i))

        while que:
            r, c = que.popleft()

            for i in range(4):
                nr, nc = r + dr[i], c + dc[i]
                if (0 <= nr < row and
                    0 <= nc < col and
                    not chk[nr][nc] and
                    board[nr][nc] == 'O'):
                    que.append((nr, nc))
                    chk[nr][nc] = True

        for i in range(1, row - 1):
            for j in range(1, col - 1):
                if board[i][j] == 'O' and not chk[i][j]:
                    board[i][j] = 'X'

File: 0130-surrounded-regions/0130-surrounded-regions.java
class Coordinates {
    int r;
    int c;

    Coordinates(int r, int c) {
        this.r = r;
        this.c = c;
    }
}

class Solution {
    public void solve(char[][] board) {
        if (board == null || board[0] == null) {
            return;
        }

        int row = board.length;
        int col = board[0].length;

        if (row < 3 || col < 3) {
            return;
        }

        int[] dr = {0, 0, -1, 1};
        int[] dc = {1, -1, 0, 0};
        boolean[][] chk = new boolean[row][col];
        Queue<Coordinates> que = new ArrayDeque<>();

        for (int i = 0; i < row; i++) {
            if (board[i][0] == 'O') {
                chk[i][0] = true;
                que.add(new Coordinates(i, 0));
            }
            if (board[i][col - 1] == 'O') {
                chk[i][col - 1] = true;
                que.add(new Coordinates(i, col - 1));
            }
        }

        for (int i = 0; i < col; i++) {
            if (board[0][i] == 'O') {
                chk[0][i] = true;
                que.add(new Coordinates(0, i));
            }
            if (board[row - 1][i] == 'O') {
                chk[row - 1][i] = true;
                que.add(new Coordinates(row - 1, i));
            }
        }

        while (!que.isEmpty()) {
            Coordinates now = que.poll();

            for (int i = 0; i < 4; i++) {
                int nr = now.r + dr[i];
                int nc = now.c + dc[i];
                if ((0 <= nr && nr < row) &&
                    (0 <= nc && nc < col) && 
                    (!chk[nr][nc] && board[nr][nc] == 'O')) {
                    que.add(new Coordinates(nr, nc));
                    chk[nr][nc] = true;
                }
            }
        }

        for (int i = 1; i < row - 1; i++) {
            for (int j = 1; j < col - 1; j++) {
                if (board[i][j] == 'O' && !chk[i][j]) {
                    board[i][j] = 'X';
                }
            }
        }
    }
}